File: 739. Daily Temperatures/739. Daily Temperatures.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        stack = []
        ans = [0] * n

        for i, temp in enumerate(temperatures):
            while stack and temp > temperatures[stack[-1]]:
                j = stack.pop()
                ans[j] = i - j
            stack.append(i)

        return ans

File: 739. Daily Temperatures/739. Daily Temperatures.java
import java.util.Stack;

class Solution {

  public int[] dailyTemperatures(int[] temperatures) {
    int n = temperatures.length;
    Stack<Integer> stack = new Stack<>();
    int[] ans = new int[n];

    for (int i = 0; i < n; i++) {
      while (!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]) {
        int j = stack.pop();
        ans[j] = i - j;
      }
      stack.push(i);
    }

    return ans;
  }
}
